### PR TITLE
[CI] code coverage by grcov

### PR DESCRIPTION
**Pull request by sausagee**
## Motivation
We need to track the code coverage to help us ensure the basic testing
is in place to exercise all the code.

We are adding grcov to CI pipeline to collect and aggregate code
coverage information. The coverage result is uploaded to codecov.io,
which can display the aggregate and integrate the result into code
review process.

Read more about grcov at https://github.com/mozilla/grcov

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Canarying on Circle with swap space to get around linker OOM.

### CODE REVIEW

**Comment by huitseeker** [comment]
Why not use the existing script? review https://github.com/libra/libra/pull/1279 ?
We have [previously worked on adding a batch mode](https://github.com/libra/libra/pull/316#pullrequestreview-266875704) to this script for CI.

**Comment by sausagee** [comment]
Thanks for the heads up. Bob has pointed me to the script.  This draft PR is trying to get it to work on Circle, canarying with different setup on Circle.

File: .circleci/config.yml [hunk 1]
```diff
@@ -109,6 +109,46 @@ jobs:
...
+          name: Install grcov
+          command: cargo install --force grcov
+      - run:
+          name: Setup rustc flags
+          command: |
+            echo "export CARGO_INCREMENTAL=0" >> $BASH_ENV
+            echo "export RUSTFLAGS='-Zprofile -Ccodegen-units=1 -Cinline-threshold=0 -Clink-dead-code -Coverflow-checks=off -Zno-landing-pads'" >> $BASH_ENV
+      - run:
+          name: Cargo test
+          command: cargo +nightly test
```
Comment (huitseeker) [inline .circleci/config.yml]: I haven't managed to override the root `rust-toolchain` file's contents like that. The observable is that beta is picked, with the following error message:
```
Caused by:
  process didn't exit successfully: `sccache rustc - --crate-name ___ --print=file-names -Zprofile -Ccodegen-units=1 -Cinline-threshold=0 -Coverflow-checks=off -Zno-landing-pads -Cpasses=insert-gcov-profiling --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=cfg` (exit code: 1)
--- stderr
error: the option `Z` is only accepted on the nightly compiler
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -109,6 +109,46 @@ jobs:
...
+      - run:
+          name: Generate coverage report
+          command: |
+            zip -0 ccov.zip `find . \( -name "libra*.gc*" \) -print`
```
Comment (huitseeker) [inline .circleci/config.yml]: You might be missing modules?
```
zip -0 ccov.zip $(find . \( -name "*.gc*" \) -print);
```

File: .circleci/config.yml [hunk 3]
```diff
@@ -109,6 +109,46 @@ jobs:
...
+            grcov ccov.zip -s . -t lcov --llvm --branch --ignore-not-existing --ignore "/*" -o lcov.info
```
Comment (huitseeker) [inline .circleci/config.yml]: Somewhat concerned about `--ignore-not-existing`: we've never needed it so far?

File: .circleci/config.yml (hunk 1 above)
Comment (sausagee) [reply .circleci/config.yml]: I got the hint from here https://github.com/rust-lang/rustup.rs#working-with-distribution-rust-packages

File: .circleci/config.yml (hunk 3 above)
Comment (sausagee) [reply .circleci/config.yml]: Going to move that to #1420 and continue

**Comment by sausagee** [comment]
This PR is an experiment splitting off from #1420 to test Circle's machine executor setup. 

If we build in Circle's primary container, linker is constantly killed due to oom. The primary container doesn't have swap.  It doesn't allow one to create their own swap file either.
```
  = note: /usr/bin/ld: final link failed: Cannot allocate memory
          collect2: error: ld returned 1 exit status
```

The linker works on `machine` executor class on circle and didn't get killed (https://circleci.com/gh/libra/libra/12328?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link). However it's awfully slow and makes latency sensitive test flaky.

I am closing this PR for now. Will continue grcov discussion in #1420.

**Pull request closed without merging**